### PR TITLE
Bond option pricing

### PR DESCRIPTION
**Pull request by marc-henrard**
* Based on black formula and yield volatility. 
* Pricer: pv and sensitivities.
* Corrected a bug in bond forward price sensitivity

### CODE REVIEW

**Review by yukiiwashita — CHANGES_REQUESTED** [review]

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingFixedCouponBondProductPricer.java [hunk 1]
```diff
@@ -656,6 +785,57 @@ public double modifiedDurationFromYield(ResolvedFixedCouponBond bond, LocalDate
...
+  public ValueDerivatives modifiedDurationFromYieldAd(ResolvedFixedCouponBond bond, LocalDate settlementDate, double yield) {
+    ImmutableList<FixedCouponBondPaymentPeriod> payments = bond.getPeriodicPayments();
+    int nCoupon = payments.size() - couponIndex(payments, settlementDate);
+    FixedCouponBondYieldConvention yieldConv = bond.getYieldConvention();
+    if (nCoupon == 1) {
+      if (yieldConv.equals(US_STREET) || yieldConv.equals(DE_BONDS)) {
+        double couponPerYear = bond.getFrequency().eventsPerYear();
+        double factor = factorToNextCoupon(bond, settlementDate);
+        double md = factor / couponPerYear / (1d + factor * yield / couponPerYear);
+        double yieldBar = factor / couponPerYear /
```
[the remark this answers is not in the document]
Comment (marc-henrard) [reply modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingFixedCouponBondProductPricer.java:811]: Corrected. Added the missing relevant test.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingFixedCouponBondProductPricer.java [hunk 2]
```diff
@@ -656,6 +785,57 @@ public double modifiedDurationFromYield(ResolvedFixedCouponBond bond, LocalDate
...
+        double md = 0d;
+        double yieldBar = 0.0;
+        return ValueDerivatives.of(md, DoubleArray.of(yieldBar));
+      }
+      double maturity = bond.getDayCount().relativeYearFraction(settlementDate, maturityDate);
+      double num = 1d + bond.getFixedRate() * maturity;
+      double den = 1d + yield * maturity;
+      double dirtyPrice = dirtyPriceFromCleanPrice(bond, settlementDate, num / den);
+      double md = num * maturity / (den * den) / dirtyPrice;
+      double denBar = -2.0d * num * maturity / (den * den * den) / dirtyPrice;
```
[the remark this answers is not in the document]
Comment (marc-henrard) [reply modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingFixedCouponBondProductPricer.java:831]: Corrected. Added the missing relevant test.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/NormalBondYieldExpiryDurationVolatilities.java [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+   * <p>
+   * The volatilities are calibrated for this date-time.
+   */
+  @PropertyDefinition(validate = "notNull", overrideGet = true)
+  private final ZonedDateTime valuationDateTime;
+  /**
+   * The normal volatility surface.
+   * <p>
+   * The x-value of the surface is the expiry, as a year fraction.
+   * The y-value of the surface is the swap tenor, as a year fraction rounded to the month.
```
[the remark this answers is not in the document]
Comment (marc-henrard) [reply modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/NormalBondYieldExpiryDurationVolatilities.java:72]: Done

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/BondVolatilitiesName.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+import java.io.Serializable;
+
+import org.joda.convert.FromString;
+
+import com.opengamma.strata.collect.ArgChecker;
+import com.opengamma.strata.data.MarketDataName;
+
+/**
+ * The name of a set of swaption volatilities.
```
[the remark this answers is not in the document]
Comment (marc-henrard) [reply modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/BondVolatilitiesName.java:16]: Done

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/BondYieldSensitivity.java [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+  @PropertyDefinition(overrideGet = true)
+  private final double sensitivity;
+
+  //-------------------------------------------------------------------------
+  /**
+   * Obtains an instance from the specified elements.
+   * 
+   * @param volatilitiesName  the name of the volatilities
+   * @param expiry  the time to expiry of the option as a year fraction
+   * @param tenor  the underlying swap tenor
```
[the remark this answers is not in the document]
Comment (marc-henrard) [reply modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/BondYieldSensitivity.java:84]: Done

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingFixedCouponBondProductPricer.java [hunk 6]
```diff
@@ -422,16 +424,19 @@ public PointSensitivityBuilder dirtyPriceSensitivity(
...
-    RepoCurveZeroRateSensitivity dfSensi = repoDf.zeroRatePointSensitivity(referenceDate)
-        .multipliedBy(-pv.getAmount() / df / df / notional);
-    return pvSensi.combinedWith(dfSensi);
+    CurrencyAmount pv = presentValue(bond, provider, settlementDate);
+    RepoCurveDiscountFactors repoDf = repoCurveDf(bond, provider);
+    double df = repoDf.discountFactor(settlementDate);
+    // Backward sweep
+    double priceBar = 1.0;
+    double pvBar = 1.0d / df / notional * priceBar;
+    double dfBar = -pv.getAmount() / (df * df) / notional * priceBar;
```
[the remark this answers is not in the document]
Comment (marc-henrard) [reply modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingFixedCouponBondProductPricer.java:436]: Yes needed. df is used in price computation. As priceBar=1, we could remove it, but I prefer to leave it  as it explain the code better to my point of view.

**Review by yukiiwashita — APPROVED** [review]

**Pull request merged**